### ISSUE TITLE
Problems with clicking on mobile

### ISSUE THREAD

**Issue by abelptvts** [transcriptic/react-map-interaction#31]
Hey there!

There are certain issues with clicking items other than MapInteraction when using this component on mobile. Think about this like there is a header with buttons above a MapInteraction component and the buttons in the header don't respond to clicking. 

The problem seems to be that the  _touchstart_, _touchmove_ and _touchend_ event listeners are set on the  **window** object rather than the **containerNode**  ref, and `e.preventDefault()` sort of "eats" the event and prevents the buttons form handling it. Setting the event listeners on the containerNode seemingly solves the problem.

I really think that it should be an option whether to set the event listeners on the window or the containerNode. Gonna submit a PR for this later.

**Comment by scottyantipa** [comment]
@abelptvts I believe this issue was fixed by #34 

I've tried to reproduce the issue on master and haven't been able to. I added a button above the component and tested clicking on it in the simulator. See below. Let me know if you are seeing a separate issue, or if the latest version (1.3.1) still doesn't work for you.

![touch-header](https://user-images.githubusercontent.com/3643611/62148245-2ec18980-b2ae-11e9-9e38-c9adee550777.gif)

I tested by just adding this to `example/App.js`
```
        <div>
          <button
            onClick={() => alert('click')}
            onTouchEnd={() => alert("onTouchEnd")}
          >
            test
          </button>
        </div>
        <div style={style}>
          <MapInteractionCSS
```

**Comment by wallzero** [comment]
Hey @scottyantipa I am experiencing this issue as well. I thought #20 was the root cause. However, even after #34 this issue persists.

For myself only `onClick` events are not activated. I've tested this on Chrome and Firefox on Android and exhibit the same bug. I can also replicate the issue on my desktop in Chrome with developer tools and mobile device mode. I _cannot_ replicate the issue in Firefox on desktop with developer tools and mobile device mode.

I have an experimental graph library I am working on. You can see in [this demo](https://wallzero.gitlab.io/react-ag/#/en/ComplexDiagram) the buttons no longer respond.

**Comment by abelptvts** [comment]
Hey, @wallzero, would you mind trying it with my PR? I'm just really curious :) (it's my first public PR, it would feel really good if I solved something :) )

**Comment by wallzero** [comment]
@abelptvts I cloned your fork and tried it out with `yarn link`. It works, but, panning stops once my mouse leaves the element. For example, [`react-draggable`](https://mzabriskie.github.io/react-draggable/example/) also continues dragging so long as the mouse/finger is down. In my use case, until the mouse/finger is lifted, panning shouldn't stop regardless of whether it leaves the container - and this is the current default behaviour of `react-map-interaction`. 

While I don't think it is a solution to this issue, it could be a useful additional feature for other use cases.

**Comment by zlanich** [comment]
I can also confirm that both things mentioned here are problems in my app as well. Details:

**Issue 1: Touch End does not fire if you move your finger and TouchMove occurs.** 

This is eating my `touchend` events that I have attached to the container and making it impossible for me to rely on a `touchend` event if the user moves their finger. The reason it's an issue with my app is because I have a special `<TapOrClick />` component that simulates "tap" events on an element by attaching my handler on `touchend` and preventing the handler from being called if the user moves their finger too far. 

Screencast of logged behavior: http://zsl.io/Ws2XsG (see that the `touchend` event does not fire after `touchmove`).

**Issue 2: Panning abruptly stops if you hit any bounds. This is a huge issue, especially when the map is zoomed all the way out.** 

Screencast: http://zsl.io/H954th

NOTE: I have special bounds needs (similarly mentioned in #10), so rather than using the min/max scale props, I'm simply "clamping" the scale/translation coming out of `onChange` before calling `this.setState({  scale, translation });`.

Since I have this working, I may be interested in helping on #10, but I'm also doing this on a special use-case and measuring the pan-able image on load as my base numbers for the clamping math, so I don't know if it would make sense to work that into this library or not.

Please help! (and thank you, of course)

**Comment by scottyantipa** [comment]
@zlanich Thanks for the super detailed reports, I appreciate the screencasts.

> Issue 1: Touch End does not fire if you move your finger and TouchMove occurs.
This is, somewhat by design. When the user is touches down on elementA on the canvas, and then drags, and then touches up, we dont want that to trigger a click event on elementA. We only trigger the click event on elementA if its a normal click action (down and immediately up). This means that the TouchEnd event doesnt bubble up. However, if what you need to do is listen for MapInteraction to fire a TouchEnd event anytime the user stops dragging, I think we could put in a prop for that. Is this what you're looking for?

> Issue 2: Panning abruptly stops if you hit any bounds. This is a huge issue, especially when the map is zoomed all the way out.
Hm, Im not sure from the screencast if this is expected behavior based on the clamp you are setting, or if there is something wrong with MapInteraction. You *would* expect the abrupt stop if you are clamping the translation (there isn't a animation or spring effect).

> Issue 3: The map is interfering with all native mobile events in the entire app, so I can't click in to text boxes, click on hyperlinks, etc on mobile. Works on desktop, though.
You're correct, I hadn't encountered this before. I repro'd and am working on a fix.

I put up a branch which adds storybook and also comments out a few of the offending lines from issue 3, if you're curious to see if it solves your problem. I'll put up a PR soon with storybook and a fix.
https://github.com/transcriptic/react-map-interaction/tree/storybook

**Comment by scottyantipa** [comment]
Ok a fixed is released in 1.3.2

**Comment by SafirMonroe** [comment]
The issue persists for me if i try to use an onclick function for an element inside the MapInteractionCSS tag in version 2.1.0.  As a fix, I used onTouchEnd instead and it worked.
